File: modules/mod_huff.R
# Módulo para cálculo de Huff entre sucursales y competencia

source("utils/huff_model.R", local = TRUE)
source("utils/google_places.R", local = TRUE)

mod_huff_ui <- function(id) {
  ns <- NS(id)
  tagList(
    fluidRow(
      column(
        width = 4,
        selectInput(
          ns("sucursal"),
          "Sucursal Rosa Oliva:",
          choices = APP_CONFIG$sucursales_rosa_data$nombre # From APP_CONFIG
        ),
        actionButton(ns("calcular"), "Calcular captación", class = "btn-primary"),
        br(), br(),
        h4("Parámetros del modelo:"),
        numericInput(ns("alfa"), "Atractivo (α):",
                     value = APP_CONFIG$huff_default_alfa, # From APP_CONFIG
                     min = 0.1, max = 3, step = 0.1),
        numericInput(ns("beta"), "Fricción distancia (β):",
                     value = APP_CONFIG$huff_default_beta, # From APP_CONFIG
                     min = 0.1, max = 5, step = 0.1)
      ),
      column(
        width = 8,
        h4("Resultados de captación"),
        DT::dataTableOutput(ns("tabla_resultados")),
        br(),
        h4("Mapa de análisis"),
        leaflet::leafletOutput(ns("mapa_huff"), height = "400px")
      )
    )
  )
}

mod_huff_server <- function(id) {
  moduleServer(id, function(input, output, session) {
    ns <- session$ns
    
    if (!nzchar(Sys.getenv("GOOGLE_PLACES_API_KEY"))) {
      shiny::showNotification(
        "Advertencia: GOOGLE_PLACES_API_KEY no configurada. Atractividad de competencia usará valores por defecto.",
        type = "warning",
        duration = NULL
      )
    }
    
    # --- Reactive Data Definitions ---
    sucursales_rosa <- reactive({
      APP_CONFIG$sucursales_rosa_data # From APP_CONFIG
    })
    
    competencia_base <- reactive({
      APP_CONFIG$competencia_base_data # From APP_CONFIG
    })
    
    competencia <- reactive({
      df <- competencia_base()
      if (nrow(df) == 0) {
        return(df)
      }
      df$atractivo <- DEFAULT_RATING
      for (i in 1:nrow(df)) {
        rating <- get_google_place_rating(
          place_name = df$nombre[i],
          lat = df$lat[i],
          lng = df$lng[i]
        )
        df$atractivo[i] <- rating
      }
      return(df)
    })

    req(agebs_hex)

    agebs_data <- reactive({
      if (exists("agebs_hex") && is.data.frame(get("agebs_hex")) && nrow(get("agebs_hex")) > 0) {
        local_agebs_hex <- get("agebs_hex")

        # Ensure necessary columns for processing are present
        required_cols <- c("id_hex", "clientes_totales") # Add any other essential non-geometry columns
        missing_cols <- setdiff(required_cols, names(local_agebs_hex))
        if (length(missing_cols) > 0) {
          stop(paste("Columnas requeridas faltantes en agebs_hex:", paste(missing_cols, collapse = ", ")))
        }

        tryCatch({
          # Dynamically find the sfc geometry column
          sfc_column_indices <- which(sapply(local_agebs_hex, function(col) inherits(col, "sfc")))

          if (length(sfc_column_indices) == 0) {
            stop("No sfc geometry column found in agebs_hex.")
          }

          geom_col_name <- names(local_agebs_hex)[sfc_column_indices[1]]
          # message(paste("Using geometry column:", geom_col_name)) # For debugging

          # Additional safety check (might be redundant if the above is robust)
          if (!inherits(local_agebs_hex[[geom_col_name]], "sfc")) {
            stop(paste("Column '", geom_col_name, "' is not an sfc column despite initial detection."))
          }

          # Proceed with centroid calculation using the dynamically found geometry column
          centroids_sf <- sf::st_centroid(local_agebs_hex[[geom_col_name]])
          coords <- sf::st_coordinates(centroids_sf)

          data.frame(
            cvegeo = local_agebs_hex$id_hex, # Assuming id_hex exists
            lat = coords[, 2],
            lng = coords[, 1],
            poblacion = local_agebs_hex$clientes_totales, # Assuming clientes_totales exists
            stringsAsFactors = FALSE
          )
        }, error = function(e) {
          shiny::showNotification(paste("Error procesando agebs_hex:", e$message), type = "error")
          message(paste("Detailed error in agebs_data reactive:", e)) # More detailed console log
          # Return an empty data.frame with the expected structure in case of error
          data.frame(
            cvegeo = character(), lat = numeric(), lng = numeric(),
            poblacion = numeric(), stringsAsFactors = FALSE
          )
        })
      } else {
        shiny::showNotification("Usando datos AGEB simulados (agebs_hex no disponible o vacío).", type = "message")
        # Fallback to simulated data if agebs_hex is not available
        data.frame(
          cvegeo = paste0("AGBSIM", 1:20),
          lat = runif(20, 17.04, 17.08),
          lng = runif(20, -96.74, -96.70),
          poblacion = sample(200:1500, 20, replace = TRUE),
          stringsAsFactors = FALSE
        )
      }
    })
    
    # --- Calculation Logic ---
    resultados <- eventReactive(input$calcular, {
      req(input$sucursal, agebs_data(), competencia())
      
      if (nrow(agebs_data()) == 0) {
        shiny::showNotification("No hay datos de AGEBs disponibles para el cálculo.", type = "error")
        return(data.frame(Error = "Datos de AGEBs no disponibles.", stringsAsFactors = FALSE))
      }
      
      id_calc_notif <- shiny::showNotification("Calculando captación...", type = "message", duration = NULL)
      on.exit(shiny::removeNotification(id_calc_notif), add = TRUE)
      
      tryCatch({
        suc_sel <- sucursales_rosa() %>% dplyr::filter(nombre == input$sucursal)
        if (nrow(suc_sel) == 0) {
          stop("Sucursal seleccionada no encontrada.")
        }
        
        comp_data <- competencia()
        todos_puntos <- dplyr::bind_rows(
          suc_sel %>% dplyr::mutate(tipo = "Rosa Oliva"),
          comp_data %>% dplyr::mutate(tipo = "Competencia")
        )
        
        current_agebs <- agebs_data()
        
        resultados_list <- lapply(1:nrow(current_agebs), function(i) {
          res_huff <- huff_model(
            ag_lat = current_agebs$lat[i],
            ag_lng = current_agebs$lng[i],
            puntos = todos_puntos,
            alfa = input$alfa,
            beta = input$beta
          )
          data.frame(
            cvegeo = current_agebs$cvegeo[i],
            poblacion = current_agebs$poblacion[i],
            agebs_lat = current_agebs$lat[i],
            agebs_lng = current_agebs$lng[i],
            res_huff,
            stringsAsFactors = FALSE
          )
        })
        resultados_detalle <- dplyr::bind_rows(resultados_list)
        
        total_poblacion_agebs <- sum(current_agebs$poblacion, na.rm = TRUE)
        if (total_poblacion_agebs == 0) {
          stop("La población total de los AGEBs es cero. No se puede calcular participación.")
        }
        
        captacion_resumen <- resultados_detalle %>%
          dplyr::group_by(id, nombre, tipo) %>%
          dplyr::summarise(
            captacion_total = sum(prob * poblacion, na.rm = TRUE),
            participacion = round((sum(prob * poblacion, na.rm = TRUE) / total_poblacion_agebs) * 100, 2),
            agebs_influencia = dplyr::n_distinct(cvegeo[prob > 0.4 & is.finite(prob)]),
            .groups = "drop"
          ) %>%
          dplyr::arrange(desc(captacion_total))
        
        session$userData$resultados_detalle <- resultados_detalle
        return(captacion_resumen)
        
      }, error = function(e) {
        shiny::showNotification(paste("Error en cálculo:", e$message), type = "error", duration = 10)
        message("Error en eventReactive resultados: ", e$message)
        return(data.frame(Error = as.character(e$message), stringsAsFactors = FALSE))
      })
    })
    
    # --- Outputs ---
    output$tabla_resultados <- DT::renderDataTable({
      res <- resultados()
      req(res)

      if ("Error" %in% names(res)) {
        DT::datatable(res[, "Error", drop = FALSE], options = list(dom = 't'))
      } else {
        DT::datatable(
          res %>% dplyr::select(
            Negocio = nombre, Tipo = tipo, `Captación estimada` = captacion_total,
            `Participación (%)` = participacion, `AGEBs influencia` = agebs_influencia
          ),
          options = list(pageLength = 5, dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel')),
          rownames = FALSE, extensions = 'Buttons'
        ) %>% DT::formatRound(c("Captación estimada"), digits = 0)
      }
    })
    
    output$mapa_huff <- leaflet::renderLeaflet({
      leaflet::leaflet() %>%
        leaflet::addProviderTiles("CartoDB.Positron") %>%
        leaflet::setView(lng = APP_CONFIG$default_lng, lat = APP_CONFIG$default_lat, zoom = 13) # Use APP_CONFIG
    })
    
    observe({
      res_data_map <- resultados()
      req(res_data_map)
      
      if (!"Error" %in% names(res_data_map) && !is.null(session$userData$resultados_detalle)) {
        detalle_map <- session$userData$resultados_detalle
        suc_sel_nombre_map <- input$sucursal
        
        ageb_perf_selected_store_map <- detalle_map %>%
          dplyr::filter(nombre == suc_sel_nombre_map & tipo == "Rosa Oliva")

        proxy <- leaflet::leafletProxy("mapa_huff", data = ageb_perf_selected_store_map)
        proxy %>% leaflet::clearMarkers() %>% leaflet::clearShapes() %>% leaflet::clearControls()
        
        if (nrow(ageb_perf_selected_store_map) > 0) {
          prob_domain <- ageb_perf_selected_store_map$prob
          if (all(is.na(prob_domain)) || length(unique(na.omit(prob_domain))) <= 1) {
            pal <- leaflet::colorNumeric("Reds", domain = c(0, 1), na.color = "transparent")
          } else {
            pal <- leaflet::colorNumeric("Reds", domain = prob_domain, na.color = "transparent")
          }
          
          proxy %>%
            leaflet::addAwesomeMarkers(
              data = sucursales_rosa(),
              lng = ~lng, lat = ~lat,
              popup = ~paste0("<b>", nombre, "</b><br>Atractivo (interno): ", atractivo),
              layerId = ~paste0("suc_", id),
              icon = leaflet::awesomeIcons(icon = "store", library = "fa", markerColor = "darkgreen")
            ) %>%
            leaflet::addAwesomeMarkers(
              data = competencia(),
              lng = ~lng, lat = ~lat,
              popup = ~paste0("<b>", nombre, "</b><br>Atractivo (Google): ", round(atractivo, 2)),
              layerId = ~paste0("comp_", id),
              icon = leaflet::awesomeIcons(icon = "store", library = "fa", markerColor = "black")
            ) %>%
            leaflet::addCircleMarkers(
              lng = ~agebs_lng, lat = ~agebs_lat,
              radius = ~ scales::rescale(sqrt(poblacion), to = c(3, 12)),
              color = ~pal(prob),
              fillOpacity = 0.6,
              stroke = TRUE, weight = 1,
              layerId = ~paste0("ageb_", cvegeo),
              popup = ~paste0(
                "<b>AGEB: ", cvegeo, "</b><br>",
                "Población (clientes potenciales): ", scales::comma(poblacion, accuracy = 1), "<br>",
                "Prob. captación (", suc_sel_nombre_map, "): ", scales::percent(prob, accuracy = 0.1)
              )
            ) %>%
            leaflet::addLegend(
              position = "bottomright",
              pal = pal,
              values = ~prob,
              title = paste0("Prob. Captación<br>", suc_sel_nombre_map),
              opacity = 1,
              layerId = "legend_huff"
            )
        }
      }
    })
  }) # End moduleServer
}

File: modules/mod_huff.py
"""
Huff model module for market capture analysis
"""

import streamlit as st
import pandas as pd
import numpy as np
import folium
from streamlit_folium import st_folium
import plotly.express as px
from pathlib import Path
import sys

# Add utils to path
sys.path.append(str(Path(__file__).parent.parent / "utils"))

from config import APP_CONFIG, GOOGLE_PLACES_API_KEY
from huff_model import huff_model
from google_places import get_google_place_rating

def render_huff_page():
    st.header("📊 Análisis de Captación - Modelo Huff")
    
    if not GOOGLE_PLACES_API_KEY:
        st.warning("⚠️ GOOGLE_PLACES_API_KEY no configurada. Atractividad de competencia usará valores por defecto.")
    
    # Check if we have map data
    map_data = st.session_state.get('map_data', {})
    
    col1, col2 = st.columns([1, 2])
    
    with col1:
        st.subheader("Configuración")
        
        # Sucursal selection
        if map_data.get('clicked_sucursal'):
            st.info("🎯 Usando ubicación seleccionada en el mapa")
            sucursal_nombre = "Sucursal Oportunidad"
            sucursales_data = pd.DataFrame({
                'id': ['clicked'],
                'nombre': ['Sucursal Oportunidad'],
                'atractivo': [100],
                'lat': [map_data['clicked_sucursal']['lat']],
                'lng': [map_data['clicked_sucursal']['lng']]
            })
        else:
            st.info("📍 Usando sucursales predefinidas")
            sucursales_data = APP_CONFIG['sucursales_rosa_data']
            sucursal_nombre = st.selectbox(
                "Sucursal Rosa Oliva:",
                sucursales_data['nombre'].tolist()
            )
        
        # Model parameters
        st.subheader("Parámetros del modelo")
        alfa = st.number_input(
            "Atractivo (α):",
            value=float(APP_CONFIG['huff_default_alfa']),
            min_value=0.1,
            max_value=3.0,
            step=0.1
        )
        
        beta = st.number_input(
            "Fricción distancia (β):",
            value=float(APP_CONFIG['huff_default_beta']),
            min_value=0.1,
            max_value=5.0,
            step=0.1
        )
        
        # Calculate button
        calcular = st.button("🔄 Calcular captación", type="primary")
    
    with col2:
        st.subheader("Resultados de captación")
        
        if calcular:
            with st.spinner("Calculando captación..."):
                resultados = calculate_huff_model(
                    sucursal_nombre, 
                    sucursales_data,
                    map_data,
                    alfa, 
                    beta
                )
                
                if resultados is not None and len(resultados) > 0:
                    # Display results table
                    display_cols = ['Negocio', 'Tipo', 'Captación estimada', 'Participación (%)', 'AGEBs influencia']
                    st.dataframe(
                        resultados[display_cols],
                        use_container_width=True,
                        hide_index=True
                    )
                    
                    # Store results for map
                    st.session_state.huff_results = resultados
                    st.session_state.selected_sucursal = sucursal_nombre
                else:
                    st.error("Error en el cálculo del modelo Huff")
        
        # Display map if we have results
        if hasattr(st.session_state, 'huff_results'):
            st.subheader("Mapa de análisis")
            huff_map = create_huff_map()
            st_folium(huff_map, width=700, height=400)

def calculate_huff_model(sucursal_nombre, sucursales_data, map_data, alfa, beta):
    """Calculate Huff model results"""
    
    try:
        # Get selected sucursal
        suc_sel = sucursales_data[sucursales_data['nombre'] == sucursal_nombre]
        if len(suc_sel) == 0:
            st.error("Sucursal seleccionada no encontrada")
            return None
        
        # Get competition data
        if map_data.get('competencia') is not None and len(map_data['competencia']) > 0:
            comp_data = map_data['competencia'].copy()
        else:
            comp_data = APP_CONFIG['competencia_base_data'].copy()
        
        # Add attractiveness to competition
        comp_data['atractivo'] = 3.5  # Default rating
        
        for idx, row in comp_data.iterrows():
            if GOOGLE_PLACES_API_KEY:
                rating = get_google_place_rating(
                    row['nombre'],
                    row['lat'] if 'lat' in row else row['latitud'],
                    row['lng'] if 'lng' in row else row['longitud']
                )
                comp_data.loc[idx, 'atractivo'] = rating
        
        # Combine all points
        suc_sel_copy = suc_sel.copy()
        suc_sel_copy['tipo'] = 'Rosa Oliva'
        
        comp_data_copy = comp_data.copy()
        comp_data_copy['tipo'] = 'Competencia'
        
        # Standardize column names
        if 'latitud' in comp_data_copy.columns:
            comp_data_copy = comp_data_copy.rename(columns={'latitud': 'lat', 'longitud': 'lng'})
        
        todos_puntos = pd.concat([suc_sel_copy, comp_data_copy], ignore_index=True)
        
        # Get AGEB data
        agebs_hex = st.session_state.agebs_hex
        
        # Calculate centroids for AGEBs
        agebs_data = []
        for idx, row in agebs_hex.iterrows():
            if pd.notna(row['geometry']):
                centroid = row['geometry'].centroid
                agebs_data.append({
                    'cvegeo': row.get('id_hex', f'ageb_{idx}'),
                    'lat': centroid.y,
                    'lng': centroid.x,
                    'poblacion': row.get('clientes_totales', 100)
                })
        
        agebs_df = pd.DataFrame(agebs_data)
        
        if len(agebs_df) == 0:
            st.error("No hay datos de AGEBs disponibles")
            return None
        
        # Calculate Huff model for each AGEB
        resultados_detalle = []
        
        for _, ageb in agebs_df.iterrows():
            res_huff = huff_model(
                ag_lat=ageb['lat'],
                ag_lng=ageb['lng'],
                puntos=todos_puntos,
                alfa=alfa,
                beta=beta
            )
            
            for _, punto in res_huff.iterrows():
                resultados_detalle.append({
                    'cvegeo': ageb['cvegeo'],
                    'poblacion': ageb['poblacion'],
                    'agebs_lat': ageb['lat'],
                    'agebs_lng': ageb['lng'],
                    'id': punto['id'],
                    'nombre': punto['nombre'],
                    'tipo': punto['tipo'],
                    'prob': punto['prob'],
                    'distancia': punto['distancia'],
                    'utilidad': punto['utilidad']
                })
        
        resultados_df = pd.DataFrame(resultados_detalle)
        
        # Calculate summary
        total_poblacion = agebs_df['poblacion'].sum()
        
        captacion_resumen = resultados_df.groupby(['id', 'nombre', 'tipo']).agg({
            'prob': lambda x: (x * resultados_df.loc[x.index, 'poblacion']).sum(),
            'cvegeo': lambda x: len(x[resultados_df.loc[x.index, 'prob'] > 0.1])
        }).reset_index()
        
        captacion_resumen.columns = ['id', 'nombre', 'tipo', 'captacion_total', 'agebs_influencia']
        captacion_resumen['participacion'] = (captacion_resumen['captacion_total'] / total_poblacion * 100).round(2)
        
        # Rename columns for display
        captacion_resumen = captacion_resumen.rename(columns={
            'nombre': 'Negocio',
            'tipo': 'Tipo',
            'captacion_total': 'Captación estimada',
            'participacion': 'Participación (%)',
            'agebs_influencia': 'AGEBs influencia'
        })
        
        captacion_resumen = captacion_resumen.sort_values('Captación estimada', ascending=False)
        
        # Store detailed results for mapping
        st.session_state.huff_detailed = resultados_df
        
        return captacion_resumen
        
    except Exception as e:
        st.error(f"Error en cálculo: {e}")
        return None

def create_huff_map():
    """Create map showing Huff model results"""
    
    # Create base map
    m = folium.Map(
        location=[APP_CONFIG['default_lat'], APP_CONFIG['default_lng']],
        zoom_start=13,
        tiles='CartoDB positron'
    )
    
    if not hasattr(st.session_state, 'huff_detailed'):
        return m
    
    detailed_results = st.session_state.huff_detailed
    selected_sucursal = st.session_state.get('selected_sucursal', '')
    
    # Filter for selected sucursal
    sucursal_results = detailed_results[
        (detailed_results['nombre'] == selected_sucursal) & 
        (detailed_results['tipo'] == 'Rosa Oliva')
    ]
    
    if len(sucursal_results) == 0:
        return m
    
    # Add AGEB circles colored by probability
    max_prob = sucursal_results['prob'].max() if len(sucursal_results) > 0 else 1
    
    for _, row in sucursal_results.iterrows():
        prob = row['prob']
        color_intensity = prob / max_prob if max_prob > 0 else 0
        
        # Color from light red to dark red
        color = f"#{int(255 - 100 * color_intensity):02x}{int(100 + 100 * color_intensity):02x}{int(100):02x}"
        
        folium.CircleMarker(
            location=[row['agebs_lat'], row['agebs_lng']],
            radius=max(3, min(12, np.sqrt(row['poblacion']) / 10)),
            color=color,
            fillColor=color,
            fillOpacity=0.6,
            popup=f"""
            <b>AGEB:</b> {row['cvegeo']}<br>
            <b>Población:</b> {row['poblacion']:.0f}<br>
            <b>Prob. captación:</b> {prob:.1%}
            """
        ).add_to(m)
    
    # Add store markers
    map_data = st.session_state.get('map_data', {})
    
    # Add Rosa Oliva stores
    if map_data.get('clicked_sucursal'):
        folium.Marker(
            location=[map_data['clicked_sucursal']['lat'], map_data['clicked_sucursal']['lng']],
            popup="Sucursal Oportunidad",
            icon=folium.Icon(color='green', icon='store')
        ).add_to(m)
    else:
        for _, store in APP_CONFIG['sucursales_rosa_data'].iterrows():
            folium.Marker(
                location=[store['lat'], store['lng']],
                popup=f"<b>{store['nombre']}</b><br>Atractivo: {store['atractivo']}",
                icon=folium.Icon(color='green', icon='store')
            ).add_to(m)
    
    # Add competition markers
    if map_data.get('competencia') is not None:
        for _, comp in map_data['competencia'].iterrows():
            lat_col = 'lat' if 'lat' in comp else 'latitud'
            lng_col = 'lng' if 'lng' in comp else 'longitud'
            
            folium.Marker(
                location=[comp[lat_col], comp[lng_col]],
                popup=f"<b>{comp['nombre']}</b><br>Competencia",
                icon=folium.Icon(color='black', icon='store')
            ).add_to(m)
    else:
        for _, comp in APP_CONFIG['competencia_base_data'].iterrows():
            folium.Marker(
                location=[comp['lat'], comp['lng']],
                popup=f"<b>{comp['nombre']}</b><br>Competencia",
                icon=folium.Icon(color='black', icon='store')
            ).add_to(m)
    
    return m